### ISSUE TITLE
Initial Load Mode

### ISSUE THREAD

**Issue by seth-shaw-unlv** [mjordan/riprap#50]
The json_api module returns _at most_ 50 items per page. The PluginFetchResourceListFromDrupal plugin accounts for this by writing the offset to fetchresourcelis.from.drupal.pager.txt so it knows where to pick up on the next run. This works as expected. 

However, I would like to run more than 50 items at a time, especially on the initial load. I can process 50 items in 2-3 minutes and I would rather not run my cron every four minutes; especially after the initial large load is done. I would rather tell it to do up to 1,000 items every hour or perhaps 4,000 early in the morning so I don't impact user experience. (Granted, only using Drupal for the Resource List won't tax it very much, Fedora will take the brunt of it.)

So, I recommend throwing another counter in PluginFetchResourceListFromDrupal that allows us to set an upper limit on the number of items to process or, perhaps, to simply run until the pagination ends for an initial load.

**Comment by mjordan** [comment]
Agree, JSON:API's limit of 50 will be inconvenient for many Islandora repos. How do you see this working in terms of implementation? A couple options I can think of:

1. For each run of Riprap, the PluginFetchResourceListFromDrupal accumulates n > 50 resource IDs in a temporary file, and that file is then used as Rirprap's input.
1. As a variation of the above, two scheduled jobs run, one that runs frequently and that fetches 50 resource IDs and accumulates them to a temp file, and the other that executes the check_fixity command on the accumulated list of IDs.

Any other ideas?

**Comment by seth-shaw-unlv** [comment]
I would add a configurable max_resources setting and then place the json_api loop inside it. That way we use the json_api pagination until max_resources is reached. I don't think we need to stash them inside a temporary file, an array of resource IDs doesn't take that much memory. Only write to disk the last next page offset we didn't actually use.

**Comment by mjordan** [comment]
@seth-shaw-unlv can you test the issue-50 branch? There is a new setting in the Islandora config file:

```
jsonapi_page_size: 50
# The number of resources to check in one Riprap run; if absent, will use
# value defined in jsonapi_page_size. Must be a multiple of number specified
# in jsonapi_page_size.
max_resources: 100
```

If you set `max_resources` to a value that is not a multiple of the value of `
jsonapi_page_size`, Riprap will log the problem and exit.

**Comment by seth-shaw-unlv** [comment]
Test with a non-multiple of the jsonapi_page_size did, indeed, throw an error:
```
# php bin/console app:riprap:check_fixity -vvv --settings=local-config.yaml
[2019-05-20 10:09:30] app.ERROR: Configuration problem with PluginFetchResourceListFromDrupal:  max_resources is not a multiple of jsonapi_page_size {"jsonapi_page_size":50,"max_resources":101} []
Configuration problem with PluginFetchResourceListFromDrupal. Please see the log for more detail.
```

I updated the setting to be a multiple of jsonapi_page_size (50x3=150) and ran the command again but it threw an error about a problematic array_merge:
```
# php bin/console app:riprap:check_fixity -vvv --settings=local-config.yaml

In PluginFetchResourceListFromDrupal.php line 126:
                                                       
  [ErrorException]                                     
  Warning: array_merge(): Argument #2 is not an array  
                                                       

Exception trace:
 () at /opt/riprap/src/Plugin/PluginFetchResourceListFromDrupal.php:126
 App\Plugin\PluginFetchResourceListFromDrupal->execute() at /opt/riprap/src/Command/CheckFixityCommand.php:64
 App\Command\CheckFixityCommand->execute() at /opt/riprap/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /opt/riprap/vendor/symfony/console/Application.php:926
 Symfony\Component\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:89
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/console/Application.php:269
 Symfony\Component\Console\Application->doRun() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:75
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /opt/riprap/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /opt/riprap/bin/console:39
```

I threw in a `print_r($node_list_from_jsonapi);` just before that line and ran it again. The data was there and it didn't complain this time (although I wasn't sure why). This time, however, it choked on something else:
```
In PluginFetchResourceListFromDrupal.php line 158:
                                                                                       
  [ErrorException]                                                                     
  Warning: count(): Parameter must be an array or an object that implements Countable  
                                                                                       

Exception trace:
 () at /opt/riprap/src/Plugin/PluginFetchResourceListFromDrupal.php:158
 App\Plugin\PluginFetchResourceListFromDrupal->execute() at /opt/riprap/src/Command/CheckFixityCommand.php:64
 App\Command\CheckFixityCommand->execute() at /opt/riprap/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /opt/riprap/vendor/symfony/console/Application.php:926
 Symfony\Component\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:89
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/console/Application.php:269
 Symfony\Component\Console\Application->doRun() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:75
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /opt/riprap/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /opt/riprap/bin/console:39
```

Apparently `$media_list = json_decode($media_list, true);` didn't give an array back like we expected it too. However, throwing another print_r shows that, at least for a few items, we are getting the media list we want. 

I wondered if this is due to my large-scale test instance being sluggish and the script attempting to json_decode the content before it has everything (causing $media_list, etc. to return NULL which, of course, is not an array nor is it countable). 

I added some error checking and continue statements:
```
# git diff src/Plugin/PluginFetchResourceListFromDrupal.php 
diff --git a/src/Plugin/PluginFetchResourceListFromDrupal.php b/src/Plugin/PluginFetchResourceListFromDrupal.php
index 4bd5a04..8ee389d 100644
--- a/src/Plugin/PluginFetchResourceListFromDrupal.php
+++ b/src/Plugin/PluginFetchResourceListFromDrupal.php
@@ -121,7 +121,19 @@ class PluginFetchResourceListFromDrupal extends AbstractFetchResourceListPlugin
             $status_code = $response->getStatusCode();
             $node_list_from_jsonapi_json = (string) $response->getBody();
             $node_list_from_jsonapi = json_decode($node_list_from_jsonapi_json, true);
-
+            if ($node_list_from_jsonapi === NULL) {
+              if ($this->logger) {
+                  $this->logger->info(
+                      "PluginFetchResourceListFromDrupal failed to read jsonapi nodes call.",
+                      array(
+                          'page offset' => $page_offset,
+                          'page limit' => $this->page_size,
+                          'sort' => '-changed',
+                      )
+                  );
+              }
+              continue;
+            }
             if ($status_code === 200) {
                 $whole_node_list['data'] = array_merge($whole_node_list['data'], $node_list_from_jsonapi['data']);
                 $this->setPageOffset($page_offset, $node_list_from_jsonapi['links']);
@@ -154,6 +166,17 @@ class PluginFetchResourceListFromDrupal extends AbstractFetchResourceListPlugin
             $media_list = (string) $media_response->getBody();
             $media_list = json_decode($media_list, true);
 
+            if ($media_list === NULL) {
+                if ($this->logger) {
+                    $this->logger->info(
+                        "PluginFetchResourceListFromDrupal is skipping node with a failed media list jsonapi call.", 
+                        array(  
+                            'Node ID' => $nid 
+                        )       
+                    );      
+                }       
+                continue;
+            }
             if (count($media_list) === 0) {
                 if ($this->logger) {
                     $this->logger->info(
```
Adding those lines showed that, yes, I am having some intermittent failures (to be explored more) but it also got us to the end of the loops where another error popped up:
```
...
[2019-05-20 10:55:11] app.INFO: PluginFetchResourceListFromDrupal executed [] []

In AbstractSQLiteDriver.php line 43:
                                                                                                                                                                                                                                                               
  [Doctrine\DBAL\Exception\TableNotFoundException]                                                                                                                                                                                                             
  An exception occurred while executing 'SELECT f0_.id AS id_0, f0_.event_uuid AS event_uuid_1, f0_.event_type AS event_type_2, f0_.resource_id AS resource_id_3, f0_.timestamp AS timestamp_4, f0_.digest_algorithm AS digest_algorithm_5, f0_.digest_value   
  AS digest_value_6, f0_.event_detail AS event_detail_7, f0_.event_outcome AS event_outcome_8, f0_.event_outcome_detail_note AS event_outcome_detail_note_9 FROM fixity_check_event f0_ WHERE f0_.resource_id = ? AND f0_.digest_algorithm = ? AND f0_.event_  
  outcome = ? ORDER BY f0_.timestamp DESC' with params ["http:\/\/localhost:8080\/fcrepo\/rest\/masters\/ent\/ent001057-009.tif", "sha256", "success"]:                                                                                                        
                                                                                                                                                                                                                                                               
  SQLSTATE[HY000]: General error: 1 no such table: fixity_check_event                                                                                                                                                                                          
                                                                                                                                                                                                                                                               

Exception trace:
 () at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php:43
 Doctrine\DBAL\Driver\AbstractSQLiteDriver->convertException() at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:169
 Doctrine\DBAL\DBALException::wrapException() at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:145
 Doctrine\DBAL\DBALException::driverExceptionDuringQuery() at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:911
 Doctrine\DBAL\Connection->executeQuery() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/Query/Exec/SingleSelectExecutor.php:50
 Doctrine\ORM\Query\Exec\SingleSelectExecutor->execute() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:334
 Doctrine\ORM\Query->_doExecute() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:967
 Doctrine\ORM\AbstractQuery->executeIgnoreQueryCache() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:922
 Doctrine\ORM\AbstractQuery->execute() at /opt/riprap/src/Repository/FixityCheckEventRepository.php:49
 App\Repository\FixityCheckEventRepository->findReferenceFixityCheckEvent() at /opt/riprap/src/Plugin/PluginPersistToDatabase.php:16
 App\Plugin\PluginPersistToDatabase->getReferenceEvent() at /opt/riprap/src/Command/CheckFixityCommand.php:100
 App\Command\CheckFixityCommand->execute() at /opt/riprap/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /opt/riprap/vendor/symfony/console/Application.php:926
 Symfony\Component\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:89
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/console/Application.php:269
 Symfony\Component\Console\Application->doRun() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:75
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /opt/riprap/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /opt/riprap/bin/console:39

In PDOConnection.php line 63:
                                                                       
  [Doctrine\DBAL\Driver\PDOException (HY000)]                          
  SQLSTATE[HY000]: General error: 1 no such table: fixity_check_event  
                                                                       

Exception trace:
 () at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:63
 Doctrine\DBAL\Driver\PDOConnection->prepare() at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:900
 Doctrine\DBAL\Connection->executeQuery() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/Query/Exec/SingleSelectExecutor.php:50
 Doctrine\ORM\Query\Exec\SingleSelectExecutor->execute() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:334
 Doctrine\ORM\Query->_doExecute() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:967
 Doctrine\ORM\AbstractQuery->executeIgnoreQueryCache() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:922
 Doctrine\ORM\AbstractQuery->execute() at /opt/riprap/src/Repository/FixityCheckEventRepository.php:49
 App\Repository\FixityCheckEventRepository->findReferenceFixityCheckEvent() at /opt/riprap/src/Plugin/PluginPersistToDatabase.php:16
 App\Plugin\PluginPersistToDatabase->getReferenceEvent() at /opt/riprap/src/Command/CheckFixityCommand.php:100
 App\Command\CheckFixityCommand->execute() at /opt/riprap/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /opt/riprap/vendor/symfony/console/Application.php:926
 Symfony\Component\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:89
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/console/Application.php:269
 Symfony\Component\Console\Application->doRun() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:75
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /opt/riprap/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /opt/riprap/bin/console:39

In PDOConnection.php line 61:
                                                                       
  [PDOException (HY000)]                                               
  SQLSTATE[HY000]: General error: 1 no such table: fixity_check_event  
                                                                       

Exception trace:
 () at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:61
 PDO->prepare() at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:61
 Doctrine\DBAL\Driver\PDOConnection->prepare() at /opt/riprap/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:900
 Doctrine\DBAL\Connection->executeQuery() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/Query/Exec/SingleSelectExecutor.php:50
 Doctrine\ORM\Query\Exec\SingleSelectExecutor->execute() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/Query.php:334
 Doctrine\ORM\Query->_doExecute() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:967
 Doctrine\ORM\AbstractQuery->executeIgnoreQueryCache() at /opt/riprap/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php:922
 Doctrine\ORM\AbstractQuery->execute() at /opt/riprap/src/Repository/FixityCheckEventRepository.php:49
 App\Repository\FixityCheckEventRepository->findReferenceFixityCheckEvent() at /opt/riprap/src/Plugin/PluginPersistToDatabase.php:16
 App\Plugin\PluginPersistToDatabase->getReferenceEvent() at /opt/riprap/src/Command/CheckFixityCommand.php:100
 App\Command\CheckFixityCommand->execute() at /opt/riprap/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /opt/riprap/vendor/symfony/console/Application.php:926
 Symfony\Component\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:89
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /opt/riprap/vendor/symfony/console/Application.php:269
 Symfony\Component\Console\Application->doRun() at /opt/riprap/vendor/symfony/framework-bundle/Console/Application.php:75
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /opt/riprap/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /opt/riprap/bin/console:39
```

However, the table does exist and has data. I haven't changed any of the database configuration, either. So, I'm not sure what is going on there.

**Comment by seth-shaw-unlv** [comment]
Turns out that when I switched branches I had stashed my .env changes with the database settings. So, once I fixed those I ran it again (with my error checking in place) and the max_resources set to 2000. It is still running, but the resource list plugin completed successfully:
```
# time php bin/console app:riprap:check_fixity -vvv --settings=local-config.yaml
[2019-05-20 12:17:03] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:09] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:14] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:19] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:24] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:29] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:34] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:39] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:44] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:49] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:17:55] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:00] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:06] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:11] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:15] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:20] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:26] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:30] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:35] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:41] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:45] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:50] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:18:56] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:00] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:05] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:10] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:15] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:20] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:25] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:30] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:35] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:41] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:46] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:50] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:19:56] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:20:00] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:20:05] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:20:10] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:20:15] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:20:20] app.INFO: PluginFetchResourceListFromDrupal is skipping node with an empty media list. {"Node ID":233854} []
[2019-05-20 12:20:21] app.INFO: PluginFetchResourceListFromDrupal executed [] []
```

I'll report back again once this run completes.

**Comment by seth-shaw-unlv** [comment]
The run completed successfully:
```
Riprap checked 1960 resources (1960 successful events, 0 failed events) in 7154.351 seconds.

real	119m16.941s
user	0m24.896s
sys	0m2.439s
```

**Comment by mjordan** [comment]
@seth-shaw-unlv cool, thanks, I'll merge the issue-50 branch.

**Comment by mjordan** [comment]
Resolved with 3071c1c1224b3627109fab6d0228e78528e363eb.
